lib/gc/Transforms/DeepTileContractionNamedOp.cpp

```diff
@@ -0,0 +1,1032 @@
+  Location loc = linalgOp->getLoc();
+  DtypeLegalizeResult result;
+  if (!linalgOp)
+    return failure();
+
+  if (needToLegalizeDtype(linalgOp)) {
+    rewriter.setInsertionPoint(linalgOp);
+    IRMapping mapping;
+    Operation *initOp = linalgOp.getDpsInits()[0].getDefiningOp();
+    Value initValue = initOp->getResult(0);
+    ShapedType initType = cast<ShapedType>(initValue.getType());
+    ArrayRef<int64_t> tensorShape = initType.getShape();
+    SmallVector<OpFoldResult> mixedShape;
+    for (size_t i = 0UL; i < tensorShape.size(); i++) {
+      if (initType.isDynamicDim(i)) {
+        Value val = rewriter.create<tensor::DimOp>(loc, initValue, i);
+        mixedShape.push_back(val);
+      } else {
+        mixedShape.push_back(
+            getAsIndexOpFoldResult(rewriter.getContext(), tensorShape[i]));
+      }
+    }
+    Operation *currentOp;
+
+    currentOp = rewriter.create<tensor::EmptyOp>(
+        loc, mixedShape, Float32Type::get(op->getContext()));
+    if (needCopyInit) {
+      currentOp = rewriter.create<linalg::CopyOp>(loc, initOp->getResult(0),
+                                                  currentOp->getResult(0));
+    }
+    SmallVector<Value> newOperands = linalgOp->getOperands();
+    Value oldInit = newOperands.back();
+    newOperands.back() = currentOp->getResult(0);
+
+    SmallVector<AffineMap> indexingMaps = linalgOp.getIndexingMapsArray();
+    indexingMaps.push_back(indexingMaps.back());
+    SmallVector<NamedAttribute> attrs(linalgOp->getAttrs());
+    SmallVector<Type> types = {currentOp->getResult(0).getType()};
+    if (needFurtherFuse) {
+      NamedAttribute segmentSize = rewriter.getNamedAttr(
+          "operandSegmentSizes", rewriter.getDenseI32ArrayAttr({2, 2}));
+      for (auto &attr : attrs) {
+        if (attr.getName() == "indexing_maps")
+          attr.setValue(rewriter.getAffineMapArrayAttr(indexingMaps));
+        if (attr.getName() == "operandSegmentSizes")
+          attr.setValue(segmentSize.getValue());
+      }
+      types.push_back(oldInit.getType());
+      newOperands.push_back(oldInit);
+    }
+    OperationState state(loc, linalgOp->getName(), newOperands, types, attrs);
+    state.addRegion();
+    currentOp = rewriter.create(state);
+    buildLinalgRegion(currentOp, needFurtherFuse);
+    linalg::CopyOp castOp = rewriter.create<linalg::CopyOp>(
+        loc, currentOp->getResult(0), initOp->getResult(0));
+    result.linalgOp = currentOp;
+    result.castOp = castOp;
+  }
+
+  return result;
+}
+
+// Find the parent fill op of a value and will penetrate pack/pad ops
+static Operation *findParentFillOp(Value val) {
+  SmallVector<StringRef> skipOpList = {"tensor.pack", "tensor.pad"};
+  Operation *currentOp = val.getDefiningOp();
+  while (currentOp &&
+         llvm::find(skipOpList, currentOp->getName().getStringRef()) !=
+             skipOpList.end() &&
+         !isa<linalg::FillOp>(currentOp)) {
+    currentOp = currentOp->getResult(0).getDefiningOp();
```

Comment:
There it will goes into an endless loop, shall be `currentOp->getOperand(0).getDefiningOp()` instead?

Reply:
Thanks for pointing it out! I replace it with `getOperand`